Wikipedia/Code/WelcomeAnimationExtensions.swift

```diff
@@ -0,0 +1,753 @@
+        let anim = CABasicAnimation(keyPath: "transform")
+        anim.duration = duration
+        anim.beginTime = CACurrentMediaTime() + delay
+        anim.fillMode = kCAFillModeForwards
+        anim.removedOnCompletion = false
+        anim.toValue = NSValue(CATransform3D: transform)
+        anim.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)
+        return anim;
+    }
+    class func wmf_animationToOpacity(opacity: Double, delay: Double, duration: Double) -> CABasicAnimation {
+        let anim = CABasicAnimation(keyPath: "opacity")
+        anim.duration = duration
+        anim.beginTime = CACurrentMediaTime() + delay
+        anim.fillMode = kCAFillModeForwards
+        anim.removedOnCompletion = false
+        anim.toValue = opacity
+        anim.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)
+        return anim;
+    }
+}
+
+extension UIView{
+    func wmf_addCircles(circles: WelcomeCircle...) {
+        for circle: WelcomeCircle in circles {
+            let circleShape = circle.isDashed ? CAShapeLayer.wmf_addDashedCircle(circle, toView: self) : CAShapeLayer.wmf_addCircle(circle, toView:self)
+            circleShape.wmf_addWelcomeAnimation(circle.animation);
+        }
+    }
+    func wmf_addPluses(pluses: WelcomePlus...) {
+        for plus: WelcomePlus in pluses {
+            let plusShape = CAShapeLayer.wmf_addPlus(plus, toView:self)
+            plusShape.wmf_addWelcomeAnimation(plus.animation);
+        }
+    }
+    func wmf_addLines(lines: WelcomeLine...) {
+        for line: WelcomeLine in lines {
+            let lineShape = CAShapeLayer.wmf_addLine(line, toView:self)
+            lineShape.wmf_addWelcomeAnimation(line.animation);
+        }
+    }
+    func wmf_addBars(bars: WelcomeBar...) {
+        for bar: WelcomeBar in bars {
+            let barShape = CAShapeLayer.wmf_addBar(bar, toView:self)
+            barShape.wmf_addWelcomeAnimation(bar.animation);
+        }
+    }
+    func wmf_addImages(images: WelcomeImage...) -> [UIImageView] {
+        return images.map({ (image) -> UIImageView in
+            let imageView = UIImageView(frame: image.frame)
+            imageView.image = UIImage(named: image.name)
+            imageView.contentMode = UIViewContentMode.ScaleAspectFit;
+            imageView.layer.zPosition = image.zPosition
+            imageView.layer.wmf_addWelcomeAnimation(image.animation);
+            self.addSubview(imageView)
+            return imageView
+        })
+    }
+    
+    func wmf_configureForIntroAnimation(){
+        CATransaction.begin()
+        
+        let finalDashedCircleOpacity = 0.15
+        let finalCircleOpacity = 0.09
+        let finalPlusOpacity = 0.15
+        let finalLinesOpacity = 0.15
+        let scaleZeroXF = CATransform3DMakeScale(0, 0, 1)
+        let rightXF = CATransform3DMakeTranslation(50, 0, 0)
```

Comment:
Lots os objects / structs / methods above for creating all the items, but we still need additional configuration numbers here.

All of this should be encapsulated. Basically we are mixing levels of abstraction.

At the highest level, you should only interact with fully configured objects. In pseudo code it might look something like this:

```
view = someCircle()
animation = someAnimation()
self.view.addSubview(view)
view.startAnimation(animation)